pkg/core/typeurl/types.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package typeurl is Marshal For any type
+package typeurl
+
+import (
+	"encoding/json"
+	"path"
+	"reflect"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+)
+
+// registry store the reflectType and typeurl string map
+var registry map[reflect.Type]string
+
+// revRegistry store the typeurl string and reflectType map
+var revRegistry map[string]reflect.Type
+var lock sync.Mutex
+
+const (
+	baseTypeURL = "kubeedge.io"
+)
+
+var (
+	// WrongTypeError stand for custom type error
+	WrongTypeError = errors.Errorf("Wrong Type Error")
+)
+
+func init() {
+	registry = make(map[reflect.Type]string)
+	revRegistry = make(map[string]reflect.Type)
+	lock = sync.Mutex{}
+	// Register sting
+	RegisterType("", "string")
+	// Register boolean
+	RegisterType(true, "bool")
+	// register slice
+	RegisterType([]interface{}{}, "slice.interface")
+}
+
+// getRealType get the reflectType of v, if v is Ptr, get it`s real type
+func getRealType(v interface{}) reflect.Type {
+	t := reflect.TypeOf(v)
+	switch t.Kind() {
+	case reflect.Ptr:
+		// if kind is Ptr, get it`s real type
+		for t.Kind() == reflect.Ptr {
+			t = t.Elem()
+		}
+	case reflect.Chan, reflect.Func, reflect.Interface, reflect.UnsafePointer:
+		panic(WrongTypeError)
+	}
+	return t
+}
+
+// RegisterType is Register a type with the base url of the type
+func RegisterType(v interface{}, args ...string) {
+	t := getRealType(v)
+	param := []string{baseTypeURL}
+	param = append(param, args...)
+	url := path.Join(param...)
+
+	lock.Lock()
+	defer lock.Unlock()
+	if u, ok := registry[t]; ok {
+		panic(errors.Errorf("type %v has registerd , it's path is %v", t.String(), u))
```

Comment:
is panic needed here ? if a type already registered , can we ignore and continue ?

Reply:
@rohitsardesai83 ， i think panic is necessary， We need to make sure that the realtype(`t`) and `url` are unique.